### DISCUSSION TITLE
Custom section numbering

### DISCUSSION THREAD

**Discussion by pwseo** [quarto-dev/quarto-cli#873] [Q&A] [answered]
Is there an easy way to customize how section numbering is rendered in the documents? In a certain type of documents, I'd like for the level 3 headings to be rendered as "Article X" followed by a linebreak and then the actual section title (X being the counter). In the sidebar it would be ideal to keep the default rendering, though.
Does one have to create a new template in order to do this?

**Comment by jjallaire** [comment, marked as answer]
This isn't super straightforward right now but it's likely possible if you are willing to do some hacking. If you are *only* targeting HTML you could use CSS counters (https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Counter_Styles/Using_CSS_counters)

Another approach would be to use a pandoc filter (https://quarto.org/docs/authoring/filters.html). Note that you can arrange for your filter to run before or after quarto (where quarto will actually be doing the numbering). A key thing to know is that if you write a function to intercept headers you can actually return a list of elements from the filter (so the Header is e.g. replaced with Header,Paragraph)

**Reply by pwseo** [reply]
Thanks for the quick reply, @jjallaire
I think I will try to build a pandoc filter, then :) I'm sure in the future this will be easier to accomplish, once other, more important aspects of Quarto are finished.

**Reply by jjallaire** [reply]
Okay, great! To get you started here's a simple example. This document adds a filter that will be run *after* the quarto filters:

````markdown
---
title: "pandoc-header-filter"
number-sections: true
filters: 
  - quarto
  - article-headers.lua
---

## Header

### Other Header
````

Here's the filter source code:

```lua
function Header(el)
  if el.level == 3 then
    return pandoc.List({
      el, 
      pandoc.Para(pandoc.Str("More content"))
    })
  end
end
```

Since this filter is run after quarto you'll see that if you look inside `el.content` there is a span with the heading number (you can dump an element for debugging in Lua with `print(el)`). If you instead run your filter before quarto there will no numbering (which you might want if you are implementing a counter). You can prevent automatic numbering by quarto by either setting `number-depth` or by having your filter add the `unnumbered` class to the heading.

**Reply by ratnanil** [reply]
@pwseo: Did you ever implement your own lua filter based numbering? I would be interested in this as well!

**Comment by pwseo** [comment]
Again, thank you very much for the quick reply and example code.
